### PR TITLE
cpu/esp32: add the new API function uart_mode to periph/uart

### PR DESCRIPTION
**Pull request by gschorcht**
### Contribution description

This PR adds API function `uart_mode` according to PR #10743 to `periph/uart` low level driver of ESP32 implementation. The PR depends on some in PR #11276.

### Testing procedure

1. Compile and flash `tests/periph_uart` with enabled module `periph_uart_modecfg`
```
USEMODULE=periph_uart_modecfg make BOARD=esp32-wroom-32 -C tests/periph_uart flash
```
2. Connect GPIO9 [UART_DEV(1):RX] and GPIO10 [UART_DEV(1):TX].

3. Use the following commands and observe the UART communication with a logic analyzer
```
init 1 9600
mode 1 7 o 2
send 1 hello
mode 1 6 e 2
send 1 hello
...
```
### Issues/PRs references

Depends on PR #11276 
Implementation according to #10743

### CODE REVIEW

**Comment by yegorich** [comment]
So far we have agreed upon making `uart_mode` a feature till all platforms implement it. Could you wrap `uart_mode` into `#ifdef MODULE_PERIPH_UART_MODECFG` and add the feature to `cpu/esp32/Makefile.features`?

Other than that looks good to me.

**Comment by gschorcht** [comment]
@yegorich 
> Could you wrap uart_mode into `#ifdef MODULE_PERIPH_UART_MODECFG` and add the feature to `cpu/esp32/Makefile.features`?

I didn't embed the function in `#ifdef MODULE_PERIPH_UART_MODECFG` by intention since the driver itself uses the function for its initial configuration. So it can be considered as an internal function as long as it isn't declared as an official feature. Alternatively, I see following options:

1. Alternatively, I could define two functions, a static function _uart_mode used by the driver and a API wrapper function `uart_mode`. It would cost one additional function call.

2. Another possibility would be to add `USEMODULE += periph_uart_modecfg` by default for ESP MCUs.

Furthermore, since `uart_mode` is added to `periph/uart.h` without `#ifdef MODULE_PERIPH_UART_MODECFG`, it is already documented and someone could expect to be able to use it.

**Comment by yegorich** [comment]
I think `FEATURES_PROVIDED += periph_uart_modecfg` in `cpu/esp32/Makefile.features` should be sufficient. This way `tests/periph_uart/main.c` enables uart_mode support.

@haukepetersen, @MrKevinWeiss what do you think?

**Comment by gschorcht** [comment]
@yegorich I'm not sure whether `cpu/esp32/Makefile.features` is the right location for 
```
FEATURES_PROVIDED += periph_uart_modecfg
```
since it belongs to peripheral interfaces like `periph_uart` itself for which the features are usually defined in `boards/*`.

**Comment by gschorcht** [comment]
> I think FEATURES_PROVIDED += periph_uart_modecfg in cpu/esp32/Makefile.features should be sufficient. This way tests/periph_uart/main.c enables uart_mode support.

That's true for `test/periph_uart` but not for applications that do not explicitly ask for this feature. But module `periph_uart` low level driver is always used since `UART_DEV(0)` is used as `stdio`.

**Comment by gschorcht** [comment]
I have realized it now in that way that there is an internal function `_uart_set_mode` which is used by the driver itself and a wrapper function `uart_mode` which is embedded in `#ifdef MODULE_PERIPH_UART_MODECFG` and only enabled when module `periph_uart_modecfg` is used.

File: cpu/esp32/periph/uart.c [hunk 1]
```diff
@@ -279,63 +277,151 @@ void IRAM __uart_intr_handler (void *arg)
...
+int _uart_set_baudrate(uart_t uart, uint32_t baudrate)
 {
     DEBUG("%s uart=%d, rate=%d\n", __func__, uart, baudrate);
 
     CHECK_PARAM_RET (uart < UART_NUMOF, -1);
 
+    /* wait until TX FIFI is empty */
+    while (_uarts[uart].regs->status.txfifo_cnt != 0) { }
+
+    critical_enter();
```
Comment (MrKevinWeiss) [inline cpu/esp32/periph/uart.c]: Can you put some information on why you added the critical*, does it apply to the *_mode or is it something separate

Comment (gschorcht) [reply cpu/esp32/periph/uart.c]: It has to be ensured that the reconfiguration cannot be interrupted. Even though it is something I realized when I was testing `_uart_set_mode`, I think it has to be ensured for reconfiguring the baudrate too. That's why I added it with this PR.

**Comment by gschorcht** [comment]
@MrKevinWeiss Thank you for reviewing the PR.

> Would you be willing to split the PRs into a cleanup __uart -> _uart and the mode functionality?

Hm, ... It is only a very small naming fix to get the code compatible with the reimplementation of ESP8266 in PR #11108 which will allow code deduplication with future PRs.

Sure, I could provide a separate PR, but I didn't want to open too many different development branches which later cause a lot of merging conflicts or too many dependencies of different PRs. But, if you think that it might be better to have different PRs, I would provide a separate PR for renaming and would make the this PR dependent on the separate PR. BTW, PR #10644 requires changed names too.

**Comment by gschorcht** [comment]
@MrKevinWeiss
Finally, I decided to split the PR into  `cleanup` #11276 and `uart_mode` (this PR).

At the moment, this PR includes all commits of PR #11276 since I couldn't separate it completely due to unresolvable merge conflicts. Once, PR #11276 is merged, I will rebase this PR to remove them.

**Comment by MrKevinWeiss** [comment]
No problem, the cleanup PR should get in soon.

**Comment by MrKevinWeiss** [comment]
I guess ready for rebase now :smiley:

**Comment by gschorcht** [comment]
@MrKevinWeiss Rebased to current master :smile:

File: cpu/esp32/periph/uart.c [hunk 2]
```diff
@@ -342,3 +370,74 @@ static int _uart_set_baudrate(uart_t uart, uint32_t baudrate)
...
+    _uarts[uart].data = data_bits;
+
+    /* set number of stop bits */
+    #ifdef MCU_ESP32
+    /* workaround for hardware bug when stop bits are set to 2-bit mode. */
+    switch (stop_bits) {
+        case UART_STOP_BITS_1: _uarts[uart].regs->conf0.stop_bit_num = 1;
+                               _uarts[uart].regs->rs485_conf.dl1_en = 1;
+                               break;
+        case UART_STOP_BITS_2: _uarts[uart].regs->conf0.stop_bit_num = 1;
```
Comment (MrKevinWeiss) [inline cpu/esp32/periph/uart.c]: Should this be 2?  Either that or say not implemented.

File: cpu/esp32/periph/uart.c [hunk 3]
```diff
@@ -342,3 +370,74 @@ static int _uart_set_baudrate(uart_t uart, uint32_t baudrate)
...
+                               _uarts[uart].regs->rs485_conf.dl1_en = 1;
+                               break;
+        default: LOG_TAG_ERROR("uart", "invalid number of stop bits\n");
+                 critical_exit();
+                 return UART_NOMODE;
+    }
+    #else
```
Comment (MrKevinWeiss) [inline cpu/esp32/periph/uart.c]: Is the ifdef needed here since they are all MCU_ESP32 that use this file?

**Comment by gschorcht** [comment]
@MrKevinWeiss 
> Tested against scope.

Thanks.

> Only the stop bits seem to be an issue, everything else appeared correct.

What exactly is the issue? It worked for me.

File: cpu/esp32/periph/uart.c (hunk 2 above)
Comment (gschorcht) [reply cpu/esp32/periph/uart.c]: No, it is a known hardware bug. ` _uarts[uart].regs->conf0.stop_bit_num` has to be set to `1` if there is any stop bit. With `_uarts[uart].regs->rs485_conf.dl1_en` it is decided whether there are 1 or 2 stop bits.

**Comment by MrKevinWeiss** [comment]
> What exactly is the issue? It worked for me.

No difference between 1 stop bit and 2 stop bits, it appears that the second stop bit is not being sent (always just one stop bit).  Also in the code it appears you are assigning the same thing if you set stop bits to 1 or 2.
```c
case UART_STOP_BITS_1:
    _uarts[uart].regs->conf0.stop_bit_num = 1;
    _uarts[uart].regs->rs485_conf.dl1_en = 1;
    break;
case UART_STOP_BITS_2:
    _uarts[uart].regs->conf0.stop_bit_num = 1;
    _uarts[uart].regs->rs485_conf.dl1_en = 1;
    break;
```

**Comment by yegorich** [comment]
At least this is how ESP-IDF sets stop bits: https://github.com/espressif/esp-idf/blob/106dc05903a1691c024bb61ac7b29ca728829671/components/driver/uart.c#L139

```
esp_err_t uart_set_stop_bits(uart_port_t uart_num, uart_stop_bits_t stop_bit)
{
    UART_CHECK((uart_num < UART_NUM_MAX), "uart_num error", ESP_FAIL);
    UART_CHECK((stop_bit < UART_STOP_BITS_MAX), "stop bit error", ESP_FAIL);

    UART_ENTER_CRITICAL(&uart_spinlock[uart_num]);
    //workaround for hardware bug, when uart stop bit set as 2-bit mode.
    if (stop_bit == UART_STOP_BITS_2) {
        stop_bit = UART_STOP_BITS_1;
        UART[uart_num]->rs485_conf.dl1_en = 1;
    } else {
        UART[uart_num]->rs485_conf.dl1_en = 0;
    }
    UART[uart_num]->conf0.stop_bit_num = stop_bit;
    UART_EXIT_CRITICAL(&uart_spinlock[uart_num]);
    return ESP_OK;
}
```

**Comment by gschorcht** [comment]
> > What exactly is the issue? It worked for me.
> 
> No difference between 1 stop bit and 2 stop bits, it appears that the second stop bit is not being sent (always just one stop bit). Also in the code it appears you are assigning the same thing if you set stop bits to 1 or 2.
> 
> ```c
> case UART_STOP_BITS_1:
>     _uarts[uart].regs->conf0.stop_bit_num = 1;
>     _uarts[uart].regs->rs485_conf.dl1_en = 1;
>     break;
> case UART_STOP_BITS_2:
>     _uarts[uart].regs->conf0.stop_bit_num = 1;
>     _uarts[uart].regs->rs485_conf.dl1_en = 1;
>     break;
> ```

Ups, oh yeah. It should be
```c
case UART_STOP_BITS_1:
    _uarts[uart].regs->conf0.stop_bit_num = 1;
    _uarts[uart].regs->rs485_conf.dl1_en = 0;
    break;
```

This happened propably when I changed it back from trying to remove the hack due to the hardware bug. I thought that it is not necessary to test it again since it worked before :sunglasses:

**Comment by MrKevinWeiss** [comment]
Once it is in the HiL retesting shouldn't be a problem

**Comment by gschorcht** [comment]
@MrKevinWeiss Fixed it, but I have no hardware here for testing it again.

**Comment by MrKevinWeiss** [comment]
Leave it to me!

**Comment by gschorcht** [comment]
@MrKevinWeiss Thanks for reviewing and testing.

**Pull request merged**